packages/agents-hosting/src/transcript/fileTranscriptLogger.ts

```diff
@@ -0,0 +1,409 @@
+        return
+      }
+      await fs.unlink(file)
+    })
+  }
+
+  /**
+   * Loads a transcript from a file.
+   */
+  private async loadTranscriptAsync (transcriptFile: string): Promise<Activity[]> {
+    if (!await this.pathExists(transcriptFile)) {
+      return []
+    }
+
+    const json = await fs.readFile(transcriptFile, 'utf-8')
+    const result = JSON.parse(json)
+    return result.map(Activity.fromObject)
+  }
+
+  /**
+   * Executes a file operation with exclusive locking per file.
+   * This ensures that concurrent writes to the same transcript file are serialized.
+   */
+  private async withFileLock<T> (transcriptFile: string, operation: () => Promise<T>): Promise<T> {
+    // Get the current lock chain for this file
+    const existingLock = this._fileLocks.get(transcriptFile) ?? Promise.resolve()
+
+    // Create a new lock that waits for the existing one and then performs the operation
+    const newLock = existingLock.then(async () => {
+      return await operation()
+    }).catch(error => {
+      logger.warn('Error in write chain:', error)
+      throw error
+    })
+
+    // Update the lock chain
+    this._fileLocks.set(transcriptFile, newLock)
+
+    // Wait for this operation to complete
+    try {
+      return await newLock
+    } finally {
+      // Clean up if this was the last operation in the chain
+      if (this._fileLocks.get(transcriptFile) === newLock) {
+        this._fileLocks.delete(transcriptFile)
+      }
+    }
+  }
+
+  /**
+   * Performs the actual write operation to the transcript file.
+   */
+  private async logActivityToFile (activity: Activity, transcriptFile: string): Promise<void> {
+    const activityStr = JSON.stringify(activity)
+
+    if (!await this.pathExists(transcriptFile)) {
+      const folder = path.dirname(transcriptFile)
+      if (!await this.pathExists(folder)) {
+        await fs.mkdir(folder, { recursive: true })
+      }
+      await fs.writeFile(transcriptFile, `[${activityStr}]`, 'utf-8')
+      return
+    }
+
+    // Use file handle to append efficiently
+    const fileHandle = await fs.open(transcriptFile, 'r+')
+    try {
+      const stats = await fileHandle.stat()
+
+      // Seek to before the closing bracket
+      const position = Math.max(0, stats.size - 1)
+
+      // Write the comma, new activity, and closing bracket
+      const appendContent = `,${EOL}${activityStr}]`
+      await fileHandle.write(appendContent, position)
+      // Truncate any remaining content (in case the file had trailing data)
+      await fileHandle.truncate(position + Buffer.byteLength(appendContent))
+    } finally {
+      await fileHandle.close()
+    }
+  }
+
+  /**
+   * Updates a message in the transcript.
+   */
+  private async messageUpdateAsync (activity: Activity, transcriptFile: string): Promise<void> {
+    // Load all activities
+    const transcript = await this.loadTranscriptAsync(transcriptFile)
+
+    for (let i = 0; i < transcript.length; i++) {
```

Comment:
could use 
original = transcript.find((a)=>a.id===activity.id)
get rid of for loop and return